File: src/lib/acp.ts
import { AgentId } from "./models";

export type AcpCommandKind =
  | "plan" | "model" | "compact" | "clear" | "rewind"
  | "diff" | "permissions" | "cost" | "help" | "export"
  | "status" | "effort" | "fast" | "context" | "memory" | "recall";

export type CommandExecution = "local" | "flag-inject" | "git-local";

export interface AcpCommandDef {
  kind: AcpCommandKind;
  slash: string;
  label: string;
  description: string;
  argsHint?: string;
  execution: CommandExecution;
  supportedClis: AgentId[];
}

export interface AcpCommand {
  kind: AcpCommandKind;
  args: string[];
  rawInput: string;
}

export type AcpPickerCommandKind = Extract<AcpCommandKind, "model" | "permissions" | "effort">;

export type AcpOptionSource = "runtime" | "fallback" | "manual";

export interface AcpOptionDef {
  value: string;
  label: string;
  description?: string | null;
  source: AcpOptionSource;
}

export interface AcpOptionCatalog {
  supported: boolean;
  options: AcpOptionDef[];
  note?: string | null;
}

export interface AcpCliCapabilities {
  cliId: AgentId;
  model: AcpOptionCatalog;
  permissions: AcpOptionCatalog;
  effort: AcpOptionCatalog;
}

export interface AcpCommandResult {
  success: boolean;
  output: string;
  sideEffects: AcpSideEffect[];
}

export type AcpSideEffect =
  | { type: "modelChanged"; cliId: AgentId; model: string }
  | { type: "permissionChanged"; cliId: AgentId; mode: string }
  | { type: "effortChanged"; level: string }
  | { type: "planModeToggled"; active: boolean }
  | { type: "historyCleared" }
  | { type: "contextCompacted" }
  | { type: "conversationRewound"; removedTurns: number }
  | { type: "uiNotification"; message: string };

export interface AcpSession {
  model: Partial<Record<AgentId, string>>;
  permissionMode: Partial<Record<AgentId, string>>;
  effortLevel: string | null;
  planMode: boolean;
  fastMode: boolean;
}

export function defaultAcpSession(): AcpSession {
  return {
    model: {},
    permissionMode: {},
    effortLevel: null,
    planMode: false,
    fastMode: false,
  };
}

export const ACP_COMMANDS: AcpCommandDef[] = [
  {
    kind: "plan", slash: "/plan", label: "Plan Mode",
    description: "Toggle planning mode \u2014 CLI works in plan-only mode",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "model", slash: "/model", label: "Select Model",
    description: "Change the model for the active CLI",
    argsHint: "<model-name>",
    execution: "flag-inject", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "compact", slash: "/compact", label: "Compact Context",
    description: "Trim older conversation turns to free context space",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "clear", slash: "/clear", label: "Clear History",
    description: "Clear conversation history for all CLIs",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "rewind", slash: "/rewind", label: "Rewind",
    description: "Remove the last conversation turn and its output",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "diff", slash: "/diff", label: "Git Diff",
    description: "Show uncommitted changes in the project",
    execution: "git-local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "permissions", slash: "/permissions", label: "Permissions",
    description: "Change sandbox/permission mode for the active CLI",
    argsHint: "<mode>",
    execution: "flag-inject", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "cost", slash: "/cost", label: "Usage/Cost",
    description: "Show estimated token usage across the conversation",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "help", slash: "/help", label: "Help",
    description: "Show available commands for the active CLI",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "export", slash: "/export", label: "Export",
    description: "Export conversation history as markdown",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "status", slash: "/status", label: "Status",
    description: "Show CLI version, model, and connection info",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "effort", slash: "/effort", label: "Effort Level",
    description: "Set reasoning effort for the active CLI",
    argsHint: "[low|medium|high|max]",
    execution: "flag-inject", supportedClis: ["codex", "claude"],
  },
  {
    kind: "fast", slash: "/fast", label: "Fast Mode",
    description: "Toggle fast output mode (Claude only)",
    execution: "flag-inject", supportedClis: ["claude"],
  },
  {
    kind: "context", slash: "/context", label: "Context Usage",
    description: "Show how much context window is used per CLI",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "memory", slash: "/memory", label: "Memory",
    description: "View/edit project memory (CLAUDE.md / AGENTS.md)",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
  {
    kind: "recall", slash: "/recall", label: "Recall",
    description: "Search conversation history across all CLIs in this tab",
    argsHint: "<search-query>",
    execution: "local", supportedClis: ["codex", "claude", "gemini"],
  },
];

export function parseSlashCommand(input: string): AcpCommand | null {
  if (!input.startsWith("/")) return null;
  const parts = input.slice(1).split(/\s+/);
  const name = parts[0]?.toLowerCase();
  if (!name) return null;
  const def = ACP_COMMANDS.find((c) => c.slash === `/${name}`);
  if (!def) return null;
  return { kind: def.kind, args: parts.slice(1), rawInput: input };
}

export function getCommandsForCli(cliId: AgentId): AcpCommandDef[] {
  return ACP_COMMANDS.filter((c) => c.supportedClis.includes(cliId));
}

export function isPickerCommandKind(kind: AcpCommandKind): kind is AcpPickerCommandKind {
  return kind === "model" || kind === "permissions" || kind === "effort";
}

export function getPickerCatalog(
  capabilities: AcpCliCapabilities | null | undefined,
  kind: AcpPickerCommandKind
): AcpOptionCatalog | null {
  if (!capabilities) return null;
  return capabilities[kind];
}

export function getCommandCategory(kind: AcpCommandKind) {
  switch (kind) {
    case "plan":
    case "model":
    case "permissions":
    case "effort":
    case "fast":
    case "status":
      return "session";
    case "diff":
    case "memory":
      return "workspace";
    default:
      return "history";
  }
}

export function getCommandCategoryLabel(category: ReturnType<typeof getCommandCategory>) {
  switch (category) {
    case "session":
      return "Session Controls";
    case "workspace":
      return "Workspace Tools";
    default:
      return "History & Context";
  }
}

export function filterCommands(filter: string, cliId: AgentId): AcpCommandDef[] {
  const q = filter.toLowerCase().replace(/^\//, "");
  return ACP_COMMANDS.filter((c) => {
    const matchesFilter = !q || c.slash.slice(1).startsWith(q) || c.label.toLowerCase().includes(q);
    return matchesFilter;
  }).map((c) => ({
    ...c,
    _supported: c.supportedClis.includes(cliId),
  })).sort((a, b) => {
    // Supported commands first
    const aSupported = a.supportedClis.includes(cliId) ? 0 : 1;
    const bSupported = b.supportedClis.includes(cliId) ? 0 : 1;
    return aSupported - bSupported;
  });
}

File: src-tauri/src/acp.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AcpCommand {
    pub kind: String,
    pub args: Vec<String>,
    pub raw_input: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AcpOptionDef {
    pub value: String,
    pub label: String,
    pub description: Option<String>,
    pub source: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AcpOptionCatalog {
    pub supported: bool,
    pub options: Vec<AcpOptionDef>,
    pub note: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AcpCliCapabilities {
    pub cli_id: String,
    pub model: AcpOptionCatalog,
    pub permissions: AcpOptionCatalog,
    pub effort: AcpOptionCatalog,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AcpCommandResult {
    pub success: bool,
    pub output: String,
    pub side_effects: Vec<AcpSideEffect>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", tag = "type")]
pub enum AcpSideEffect {
    ModelChanged { cli_id: String, model: String },
    PermissionChanged { cli_id: String, mode: String },
    EffortChanged { level: String },
    PlanModeToggled { active: bool },
    HistoryCleared,
    ContextCompacted,
    ConversationRewound { removed_turns: usize },
    UiNotification { message: String },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AcpSession {
    pub model: HashMap<String, String>,
    pub permission_mode: HashMap<String, String>,
    pub effort_level: Option<String>,
    pub plan_mode: bool,
    pub fast_mode: bool,
}

impl Default for AcpSession {
    fn default() -> Self {
        Self {
            model: HashMap::new(),
            permission_mode: HashMap::new(),
            effort_level: None,
            plan_mode: false,
            fast_mode: false,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AcpCommandDef {
    pub kind: String,
    pub slash: String,
    pub label: String,
    pub description: String,
    pub args_hint: Option<String>,
    pub execution: String,
    pub supported_clis: Vec<String>,
}

pub fn command_registry() -> Vec<AcpCommandDef> {
    vec![
        AcpCommandDef {
            kind: "plan".into(),
            slash: "/plan".into(),
            label: "Plan Mode".into(),
            description: "Toggle planning mode".into(),
            args_hint: None,
            execution: "local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "model".into(),
            slash: "/model".into(),
            label: "Select Model".into(),
            description: "Change the model for the active CLI".into(),
            args_hint: Some("<model-name>".into()),
            execution: "flag-inject".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "compact".into(),
            slash: "/compact".into(),
            label: "Compact Context".into(),
            description: "Trim older conversation turns to free context space".into(),
            args_hint: None,
            execution: "local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "clear".into(),
            slash: "/clear".into(),
            label: "Clear History".into(),
            description: "Clear conversation history for all CLIs".into(),
            args_hint: None,
            execution: "local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "rewind".into(),
            slash: "/rewind".into(),
            label: "Rewind".into(),
            description: "Remove the last conversation turn".into(),
            args_hint: None,
            execution: "local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "diff".into(),
            slash: "/diff".into(),
            label: "Git Diff".into(),
            description: "Show uncommitted changes in the project".into(),
            args_hint: None,
            execution: "git-local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "permissions".into(),
            slash: "/permissions".into(),
            label: "Permissions".into(),
            description: "Change sandbox/permission mode".into(),
            args_hint: Some("<mode>".into()),
            execution: "flag-inject".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "cost".into(),
            slash: "/cost".into(),
            label: "Usage/Cost".into(),
            description: "Show estimated token usage".into(),
            args_hint: None,
            execution: "local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "help".into(),
            slash: "/help".into(),
            label: "Help".into(),
            description: "Show available commands".into(),
            args_hint: None,
            execution: "local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "export".into(),
            slash: "/export".into(),
            label: "Export".into(),
            description: "Export conversation history as markdown".into(),
            args_hint: None,
            execution: "local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "status".into(),
            slash: "/status".into(),
            label: "Status".into(),
            description: "Show CLI version, model, and connection info".into(),
            args_hint: None,
            execution: "local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "effort".into(),
            slash: "/effort".into(),
            label: "Effort Level".into(),
            description: "Set reasoning effort for the active CLI".into(),
            args_hint: Some("[low|medium|high|max]".into()),
            execution: "flag-inject".into(),
            supported_clis: vec!["codex".into(), "claude".into()],
        },
        AcpCommandDef {
            kind: "fast".into(),
            slash: "/fast".into(),
            label: "Fast Mode".into(),
            description: "Toggle fast output mode (Claude only)".into(),
            args_hint: None,
            execution: "flag-inject".into(),
            supported_clis: vec!["claude".into()],
        },
        AcpCommandDef {
            kind: "context".into(),
            slash: "/context".into(),
            label: "Context Usage".into(),
            description: "Show context window usage per CLI".into(),
            args_hint: None,
            execution: "local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
        AcpCommandDef {
            kind: "memory".into(),
            slash: "/memory".into(),
            label: "Memory".into(),
            description: "View/edit project memory files".into(),
            args_hint: None,
            execution: "local".into(),
            supported_clis: vec!["codex".into(), "claude".into(), "gemini".into()],
        },
    ]
}
